types/wlr_layer_shell.c

```diff
@@ -0,0 +1,438 @@
+	.set_anchor = layer_surface_handle_set_anchor,
+	.set_exclusive_zone = layer_surface_handle_set_exclusive_zone,
+	.set_margin = layer_surface_handle_set_margin,
+	.set_keyboard_interactivity = layer_surface_handle_set_keyboard_interactivity,
+	.get_popup = layer_surface_handle_get_popup,
+};
+
+static void layer_surface_unmap(struct wlr_layer_surface *surface) {
+	// TODO: probably need to ungrab before this event
+	wlr_signal_emit_safe(&surface->events.unmap, surface);
+
+	struct wlr_layer_surface_configure *configure, *tmp;
+	wl_list_for_each_safe(configure, tmp, &surface->configure_list, link) {
+		layer_surface_configure_destroy(configure);
+	}
+
+	surface->added = surface->configured = surface->mapped = false;
+	surface->configure_serial = 0;
+	if (surface->configure_idle) {
+		wl_event_source_remove(surface->configure_idle);
+		surface->configure_idle = NULL;
+	}
+	surface->configure_next_serial = 0;
+}
+
+static void layer_surface_destroy(struct wlr_layer_surface *surface) {
+	layer_surface_unmap(surface);
+	wlr_signal_emit_safe(&surface->events.destroy, surface);
+	wl_resource_set_user_data(surface->resource, NULL);
+	wl_list_remove(&surface->link);
+	wl_list_remove(&surface->surface_destroy_listener.link);
+	wlr_surface_set_role_committed(surface->surface, NULL, NULL);
+	free(surface);
+}
+
+static void layer_surface_resource_destroy(struct wl_resource *resource) {
+	struct wlr_layer_surface *surface =
+		layer_surface_from_resource(resource);
+	if (surface != NULL) {
+		layer_surface_destroy(surface);
+	}
+}
+
+static bool wlr_layer_surface_state_changed(struct wlr_layer_surface *surface) {
+	if (!surface->configured) {
+		return true;
+	}
+
+	struct wlr_layer_surface_state *state;
+	if (wl_list_empty(&surface->configure_list)) {
+		state = &surface->current;
+	} else {
+		struct wlr_layer_surface_configure *configure =
+			wl_container_of(surface->configure_list.prev, configure, link);
+		state = &configure->state;
+	}
+
+	bool changed = state->actual_width != surface->server_pending.actual_width
+		|| state->actual_height != surface->server_pending.actual_height;
+	return changed;
+}
+
+static void wlr_layer_surface_send_configure(void *user_data) {
+	struct wlr_layer_surface *surface = user_data;
+	surface->configure_idle = NULL;
+	struct wlr_layer_surface_configure *configure =
+		calloc(1, sizeof(struct wlr_layer_surface_configure));
+	if (configure == NULL) {
+		wl_client_post_no_memory(surface->client->client);
+		return;
+	}
+
+	wl_list_insert(surface->configure_list.prev, &configure->link);
+	configure->serial = surface->configure_next_serial;
+	configure->state.actual_width = surface->server_pending.actual_width;
+	configure->state.actual_height = surface->server_pending.actual_height;
+
+	zwlr_layer_surface_v1_send_configure(surface->resource,
+			configure->serial, configure->state.actual_width,
+			configure->state.actual_height);
+}
+
+static uint32_t wlr_layer_surface_schedule_configure(
```

Comment:
The reason we have this in xdg-shell is that there are multiple configure events. The compositor sends the role-specific configure event (and maybe others because of extensions like the decoration protocol) and then ends the configure sequence with a `xdg_surface.configure`. I don't think we need to schedule configure events for layer-shell, we can directly send it since there's only one.

Reply:
aight can do